### PR TITLE
Dont UPDATE SET twice with filecache, use UPDATE SET CASE instead

### PR DESCRIPTION
**Pull request by mrow4a**
### Description

I have noticed, that in Propagator we had to use UPDATE twice, in order to filter out scenario with unset filecache size entry. The queries were as below:

```
UPDATE `oc_filecache` SET `mtime` = GREATEST(`mtime`, :dcValue1), `etag` = :dcValue2 WHERE (`storage` = :dcValue3) AND (`path_hash` IN ('hash1', 'hash2’))

UPDATE `oc_filecache` SET `size` = `size` + :dcValue1 WHERE (`storage` = :dcValue2) AND (`path_hash` IN ('hash1', 'hash2’)) AND (`size` > '-1')"
```

However, this can be optimized to 
```
UPDATE `oc_filecache` SET `mtime` = GREATEST(`mtime`, :dcValue1), `etag` = :dcValue2, `size` = CASE WHEN `size` > '-1' THEN  `size` + :dcValue3 ELSE `size` END WHERE (`storage` = :dcValue4) AND (`path_hash` IN ('hash1', 'hash2'))
```

@DeepDiver1975 @butonic @tomneedham

### CODE REVIEW

File: lib/private/Files/Cache/Propagator.php [hunk 1]
```diff
@@ -79,23 +79,22 @@ public function propagateChange($internalPath, $time, $sizeDifference = 0) {
...
-				->andWhere($builder->expr()->gt('size', $builder->expr()->literal(-1, IQueryBuilder::PARAM_INT)));
-
-			$builder->execute();
+			// if we need to update size, only update the records with calculated size (>-1)
+			$builder->set('size', $builder->createFunction('CASE' .
+					' WHEN ' . $builder->expr()->gt('size', $builder->expr()->literal(-1, IQueryBuilder::PARAM_INT)) .
+						' THEN  `size` + ' . $builder->createNamedParameter($sizeDifference) .
+					' ELSE `size`' .
+				' END'
+			));
```
Comment (mrow4a) [inline lib/private/Files/Cache/Propagator.php]: I unfortunetelly did not find any doctrine with `set()->when()`

**Comment by mrow4a** [comment]
All checks passed

**Review by DeepDiver1975 — APPROVED** [review]

**Comment by mmattel** [comment]
Backporting?

**Comment by mrow4a** [comment]
Backport to stable10 https://github.com/owncloud/core/pull/32304

**Pull request merged**